src/services/fluence-aurora-connector/modules/fluence_aurora_connector/src/jsonrpc/register_worker.rs

```diff
@@ -0,0 +1,290 @@
+use std::convert::TryInto;
+use std::str::FromStr;
+
+use clarity::{PrivateKey, Transaction};
+use ethabi::{Function, Param, ParamType, StateMutability, Token};
+use libp2p_identity::{ParseError, PeerId};
+use marine_rs_sdk::marine;
+use thiserror::Error;
+
+use crate::chain::chain_info::ChainInfo;
+use crate::curl::send_jsonrpc;
+use crate::hex::decode_hex;
+use crate::jsonrpc::register_worker::RegisterWorkerError::{
+    InvalidPrivateKey, InvalidWorkerId, ParseDealAddr,
+};
+use crate::jsonrpc::request::RequestError;
+use crate::jsonrpc::{JsonRpcError, JsonRpcReq, JSON_RPC_VERSION};
+use crate::peer_id::serialize_peer_id;
+
+#[derive(Debug, Error)]
+pub enum RegisterWorkerError {
+    #[error("invalid worker id {1}: {0:?}")]
+    InvalidWorkerId(#[source] ParseError, &'static str),
+    #[error("error encoding function inputs: {0:?}")]
+    EncodeInput(#[from] ethabi::Error),
+    #[error("invalid deal contract addr: {0:?}")]
+    ParseDealAddr(#[source] clarity::Error),
+    #[error("invalid private key")]
+    InvalidPrivateKey(#[source] Box<dyn std::error::Error>),
+    #[error("error sending transaction to rpc: {0:?}")]
+    SendTransaction(#[from] RequestError),
+    #[error(transparent)]
+    JsonRpcError(#[from] JsonRpcError),
+}
+
+#[marine]
+pub fn register_worker(
+    pat_id: Vec<u8>,
+    worker_id: &str,
+    chain: ChainInfo,
+    deal_addr: String,
+) -> Vec<String> {
+    // get network id from rpc
+    // get gas price from rpc
+    // form tx
+    // sign
+    // send tx to rpc
+    let r: Result<_, RegisterWorkerError> = try {
+        let input = encode_call(pat_id, worker_id)?;
+        let nonce = load_nonce()?;
+        let gas_price = get_gas_price()?;
+        let endpoint = chain.api_endpoint.clone();
+        let tx = make_tx(input, chain, nonce, gas_price, deal_addr)?;
+        send_tx(tx, endpoint)?
+    };
+
+    match r {
+        Ok(_) => vec![],
+        Err(err) => vec![format!("{}", err)],
+    }
+}
+
+/// Send transaction to RPC
+fn send_tx(tx: String, api_endpoint: String) -> Result<String, RegisterWorkerError> {
+    let req = JsonRpcReq {
+        id: 0,
+        jsonrpc: JSON_RPC_VERSION.to_string(),
+        method: "eth_sendRawTransaction".to_string(),
+        params: vec![tx],
+    };
+    let response = send_jsonrpc(api_endpoint, req)?;
+    let tx_id: String = response.get_result()?;
+
+    Ok(tx_id)
+}
+
+/// Load nonce from KV
```

Comment:
> KV

do you want to add a separate KV for the connector? or receive it from the spell?